src/main/java/gov/usgs/earthquake/distribution/WebSocketClient.java

```diff
@@ -10,121 +10,105 @@
+    // try to connect to server
+    WebSocketContainer container = ContainerProvider.getWebSocketContainer();
+    container.connectToServer(this, endpoint);
+  }
+
+  public void disconnect() throws Exception {
+    if (isConnected()) {
+      session.close();
+    }
   }
 
   public void run() {
-    synchronized (sync) {
-      while (!stopThread) {
-        doReconnect = false;
-        try {
+    while (!stopThread) {
+      // try to connect
+      try {
+        synchronized(close) {
           connect();
-          if (!doReconnect) {
-            sync.wait();
-          }
-        } catch (InterruptedException e1) {
-          // likely told to stop
-          stopThread = true;
-        } catch (Exception e2) {
-          // unresolvable problem connecting; try to resolve with listener
           try {
-            listener.onConnectException(e2);
-          } catch (Exception e3) {
-            // unable to handle; quit
+            // successful connection
+            close.wait();
+          } catch (InterruptedException ie) {
+            // gracefully close, disconnect if possible
             stopThread = true;
+            disconnect();
           }
         }
-      }
-
-      // make sure we're disconnected
-      if (this.session != null && this.session.isOpen()) {
+      } catch (Exception e1) { //exception on connect
+        // let listener handle if they want
         try {
-          this.session.close();
-        } catch (IOException e) {
-          // ignore IOException; just close
+          listener.onConnectException(e1);
+        } catch (Exception e2) {
+          // kill thread on listener exception
+          break;
         }
-      }
-    }
-    listener.onDisconnect();
-  }
 
-  public void connect() throws Exception {
-    // try to connect to server
-    WebSocketContainer container = ContainerProvider.getWebSocketContainer();
-    try {
-      container.connectToServer(this, endpoint);
-    } catch (Exception e1) {
-      try {
-        this.onConnectException(e1);
-      } catch (Exception e2) {
-        throw e2;
+        // reconnect for unhandled connect exception
+        if (!stopThread && reconnect) {
+          try {
+            Thread.sleep(reconnectInterval);
+          } catch (InterruptedException ie) {
+            // gracefully close
+            stopThread = true;
+          }
+        }
       }
     }
   }
 
-  private void onConnectException(Exception e) throws Exception {
-    if (e instanceof IOException || e instanceof DeploymentException) {
-      reconnect();
-    } else {
-      throw e;
-    }
-  }
-
-  private void reconnect() {
-    try {
-      Thread.sleep(timeoutMillis);
-    } catch (InterruptedException e) {
-      // told to stop
-      stopThread = true;
-    }
-    doReconnect = true;
-  }
-
   @OnOpen
   public void onOpen(Session session) {
     this.session = session;
   }
 
   @OnClose
   public void onClose(Session session, CloseReason reason) {
```

Comment:
`onClose` should set the this.session to null; inside the synchronized block and before notify.

Reply:
done